packages/python-packages/doc-warden/warden/enforce_readme_content.py

```diff
@@ -64,20 +64,165 @@ def verify_md_readme(readme, config, section_sorting_dict):
+
+    return observed_failing_patterns
+
+# gets a distinct set of ALL patterns present in a config. This is
+# important because this allows us to precalculate which patterns a given header tag will match
+def flatten_pattern_config(patterns):
     observed_patterns = []
 
-    for header in headers:
-        observed_patterns.extend(match_regex_set(header.get_text(), patterns))
+    for pattern in patterns:
+        if isinstance(pattern, dict):
+            parent_pattern, child_patterns =  next(iter(pattern.items()))
+
+            if child_patterns:
+                observed_patterns.extend(flatten_pattern_config(child_patterns))
+            observed_patterns.extend([parent_pattern])
+        else:
+            observed_patterns.extend([pattern])
+
+    return list(set(observed_patterns))
+
+# recursive solution that walks all the rules and generates rule chains from them to test 
+# that the tree actually contains sets of headers that meet the required sections
+def recursive_header_search(header_index, patterns, parent_pattern_chain=[]):
+    unobserved_patterns = []
+
+    if patterns:
+        for pattern in patterns:
+            if isinstance(pattern, dict):
+                parent_pattern, child_patterns =  next(iter(pattern.items()))
 
-    return list(set(patterns) - set(observed_patterns))
+                if not match_regex_to_headers(header_index, parent_pattern_chain + [parent_pattern]):
+                    unobserved_patterns.append(parent_pattern_chain + [parent_pattern])
+
+                parent_chain_for_children = parent_pattern_chain + [parent_pattern]
+                unobserved_patterns.extend(recursive_header_search(header_index, child_patterns, parent_chain_for_children))
+            else:
+                if not match_regex_to_headers(header_index, parent_pattern_chain + [pattern]):
+                    unobserved_patterns.append((parent_pattern_chain + [pattern]))
+
+    return unobserved_patterns
+
+# a set of headers looks like this
+# h1
+# h2
+# h1
+# h2
+# h3
+# h1
+# any "indented" headers are children of the one above it IF the
+# one above it is at a higher header level (this is actually < in comparison)
+# result of above should be a web that looks like
+# root
+#   h1
+#      h2
+#   h1
+#      h2
+#         h3
+#   h1
+# this function examines a serial set of <h> tags and generates
+# an index that allows us to interrogate a specific header for it's containing
+# headers.
+def generate_header_index(header_constructs, patterns):
+    previous_header_level = 0
+    current_header = None
+    root = HeaderConstruct(None, None)
+    current_parent = root
+    header_index = []
+    previous_node_level = 0
+
+    for index, header in enumerate(header_constructs):
+        # evaluate the level
+        current_level = int(header.name.replace('h', ''))
+
+        # h1 < h2 == we need to traverse up
+        if current_level < current_parent.level:
+            current_parent = current_parent.parent
```

Comment:
I am a bit confused here when parsing below tree.
h1
  h2
      h3
h1

I think current approach works fine if there are 3 levels. But shouldn't this be an issue if we have 4 levels.
h1
  h2 
     h3
        h4
h1

when second h1 is processed current parent is h3. Since current level is below current parent level, it will add under  current parent's parent. right?

Reply:
I don't think so, due to how the parenting works! I'll try your exact use case.

Reply:
If it _doesn't_ work, I'll definitely correct. This is intended to be entirely generic, so I don't want to be tied to a # of levels. Thanks for the callout!

Reply:
Nah. I'm still working up the test case but this is definitely a bug.

Reply:
I think having a dict of last level number and header object might be helpful.

Reply:
Resolved! I ascend until I hit a level that's actually < the one we currently look at (aka CAN be a parent).